File: frontend/src/api/patients.ts
import { api } from "./client";
import type {
  Patient,
  PaginatedResponse,
  PatientFilters,
  DashboardStats,
} from "@/types";

export function fetchPatients(filters: PatientFilters = {}) {
  const params = new URLSearchParams();
  if (filters.search) params.set("search", filters.search);
  if (filters.status) params.set("status", filters.status);
  if (filters.sort_by) params.set("sort_by", filters.sort_by);
  if (filters.sort_order) params.set("sort_order", filters.sort_order);
  if (filters.page) params.set("page", String(filters.page));
  if (filters.page_size) params.set("page_size", String(filters.page_size));

  const qs = params.toString();
  return api.get<PaginatedResponse<Patient>>(
    `/patients${qs ? `?${qs}` : ""}`
  );
}

export function fetchPatient(id: number) {
  return api.get<Patient>(`/patients/${id}`);
}

export function createPatient(data: Omit<Patient, "id" | "created_at" | "updated_at">) {
  return api.post<Patient>("/patients", data);
}

export function updatePatient(
  id: number,
  data: Partial<Omit<Patient, "id" | "created_at" | "updated_at">>
) {
  return api.put<Patient>(`/patients/${id}`, data);
}

export function deletePatient(id: number) {
  return api.delete<void>(`/patients/${id}`);
}

export function fetchDashboardStats() {
  return api.get<DashboardStats>("/dashboard/stats");
}

File: backend/app/api/routes/patients.py
import math

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session

from app.core.db import get_db
from app.schemas.patient import (
    PatientCreate,
    PatientUpdate,
    PatientResponse,
    PaginatedPatients,
)
from app.services import patient_service

router = APIRouter(prefix="/patients", tags=["patients"])


@router.get("", response_model=PaginatedPatients)
def list_patients(
    page: int = Query(1, ge=1),
    page_size: int = Query(10, ge=1, le=100),
    search: str | None = None,
    status: str | None = None,
    sort_by: str = "last_visit",
    sort_order: str = "desc",
    db: Session = Depends(get_db),
):
    patients, total = patient_service.get_patients(
        db, page, page_size, search, status, sort_by, sort_order
    )
    return PaginatedPatients(
        items=patients,
        page=page,
        page_size=page_size,
        total=total,
        total_pages=math.ceil(total / page_size) if total else 0,
    )


@router.get("/{patient_id}", response_model=PatientResponse)
def get_patient(patient_id: int, db: Session = Depends(get_db)):
    patient = patient_service.get_patient(db, patient_id)
    if not patient:
        raise HTTPException(status_code=404, detail="Patient not found")
    return patient


@router.post("", response_model=PatientResponse, status_code=201)
def create_patient(data: PatientCreate, db: Session = Depends(get_db)):
    return patient_service.create_patient(db, data)


@router.put("/{patient_id}", response_model=PatientResponse)
def update_patient(
    patient_id: int, data: PatientUpdate, db: Session = Depends(get_db)
):
    patient = patient_service.update_patient(db, patient_id, data)
    if not patient:
        raise HTTPException(status_code=404, detail="Patient not found")
    return patient


@router.delete("/{patient_id}", status_code=204)
def delete_patient(patient_id: int, db: Session = Depends(get_db)):
    if not patient_service.delete_patient(db, patient_id):
        raise HTTPException(status_code=404, detail="Patient not found")
